File: my-folder/0015-3sum/solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 2; i++) {
            // skip duplicate
            if (i > 0 && nums[i - 1] == nums[i]) continue;
            int k =  - nums[i];
            int low = i + 1;
            int high = nums.length - 1;
            
            while (low < high) {
                if (nums[low] + nums[high] == k) {
                    
                    res.add(Arrays.asList(nums[i], nums[low], nums[high]));
                    
                    while (low < high && nums[low+1] == nums[low]) low++;
                    while (low < high && nums[high-1] == nums[high])high--;
                    
                    low++;
                    high--;
                } else if (nums[low] + nums[high] < k) {
                    low++;
                } else {
                    high--;
                }
            }
        }
        
        return res;
    }
}

File: my-folder/0015-3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        return self.nSumTarget(nums, 3, 0, 0)
    
    def nSumTarget(self, nums, n, start, target):
        sz = len(nums)
        res = []
        if n < 2 or sz < n:
            return res
        if n == 2:
            lo, hi = start, sz -1
            while lo < hi:
                sum_val = nums[lo] + nums[hi]
                left, right = nums[lo], nums[hi]
                if sum_val < target:
                    while lo < hi and nums[lo] == left:
                        lo += 1
                elif  sum_val > target:
                    while lo < hi and nums[hi] == right:
                        hi -= 1
                else:
                    res.append([left, right])
                    while lo < hi and nums[lo] == left:
                        lo += 1
                    while lo < hi and nums[hi] == right:
                        hi -= 1
        else:
            for i in range(start, sz):
                if i > start and nums[i] == nums[i - 1]:
                    continue
                sub = self.nSumTarget(nums, n-1, i+1, target - nums[i] )
                for arr in sub:
                    arr.append(nums[i])
                    res.append(arr)
        return res
